### PR TITLE
feat: transfer with hook

### PR DESCRIPTION
**Pull request by anna-carroll**
## Motivation
Integrate Connext Amarok into Nomad BridgeRouter to enable native fast liquidity.

## PR Checklist

- [ ] Added Tests
- [x] Updated Documentation
- [x] Updated CHANGELOG.md for the appropriate package

### CODE REVIEW

**Comment by odyslam** [comment]
We should definetely add tests before merging this

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 1]
```diff
@@ -36,11 +37,13 @@ contract BridgeRouter is Version0, Router {
     ITokenRegistry public tokenRegistry;
     // token transfer prefill ID => LP that pre-filled message to provide fast liquidity
     mapping(bytes32 => address) public liquidityProvider;
+    // reference to connext contract on this domain
+    IConnext public connext;
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:41]: **Note:** would _prefer_ not to store this in the router. would be ideal to have a more generalized feature - the caller passes in the token recipient + call recipient + call data 🤔

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 2]
```diff
@@ -136,54 +151,58 @@ contract BridgeRouter is Version0, Router {
...
+        // format Connext transfer message
+        bytes29 _action = BridgeMessage.formatConnextTransfer(
+            _externalId,
             _amount,
-            false
+            _detailsHash
         );
+        // send message to destination chain bridge router
+        _sendTransferMessage(_destination, _tokenId, _action);
+        // TODO: emit special event?
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:205]: TODO: event interface
- emit Send?
- emit some other event?

**Comment by odyslam** [comment]
@anna-carroll if you want to decrease the CI speed, you can rebase onto master

File: packages/contracts-bridge/contracts/BridgeRouter.sol (hunk 1 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeRouter.sol:41]: agree. this should be a user-specified message field

**Comment by prestwich** [comment]
To make review easier, please remove format-only changes to `BridgeToken`, `TokenRegistry`, and `BridgeMessage`. What caused these? These changes aren't made with a local prettier run

File: packages/contracts-bridge/contracts/BridgeMessage.sol [hunk 3]
```diff
@@ -20,7 +20,8 @@ library BridgeMessage {
         TokenId, // 1
         Message, // 2
         Transfer, // 3
-        DeprecatedFastTransfer // 4
+        DeprecatedFastTransfer, // 4
+        ConnextTransfer // 5
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/BridgeMessage.sol:24]: is this better implemented as an extension of the transfer? or as a new message type?

File: packages/contracts-bridge/contracts/BridgeMessage.sol [hunk 4]
```diff
@@ -134,9 +144,34 @@ library BridgeMessage {
...
     ) internal pure returns (bytes29) {
         return
-            abi.encodePacked(Types.Transfer, _to, _amnt, _detailsHash).ref(0).castTo(
-                uint40(Types.Transfer)
-            );
+            abi
+                .encodePacked(Types.Transfer, _to, _amnt, _detailsHash)
+                .ref(0)
+                .castTo(uint40(Types.Transfer));
+    }
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/BridgeMessage.sol:151]: format-only change

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 5]
```diff
@@ -325,7 +469,7 @@ contract BridgeRouter is Version0, Router {
     }
 
     /**
-    * @dev should be impossible to renounce ownership;
+     * @dev should be impossible to renounce ownership;
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:472]: format-only change

File: packages/contracts-bridge/contracts/BridgeToken.sol [hunk 6]
```diff
@@ -244,7 +244,7 @@ contract BridgeToken is Version0, IBridgeToken, OwnableUpgradeable, ERC20 {
     }
 
     /**
-    * @dev should be impossible to renounce ownership;
+     * @dev should be impossible to renounce ownership;
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/BridgeToken.sol:247]: format-only change

File: packages/contracts-bridge/contracts/TokenRegistry.sol [hunk 7]
```diff
@@ -278,7 +278,12 @@ contract TokenRegistry is Initializable, XAppConnectionClient, ITokenRegistry {
      * @notice Determine if token is of local origin
      * @return TRUE if token is locally originating
      */
-    function isLocalOrigin(address _token) external view override returns (bool) {
+    function isLocalOrigin(address _token)
+        external
+        view
+        override
+        returns (bool)
+    {
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/TokenRegistry.sol:286]: format-only change

File: packages/contracts-bridge/contracts/TokenRegistry.sol [hunk 8]
```diff
@@ -406,7 +411,7 @@ contract TokenRegistry is Initializable, XAppConnectionClient, ITokenRegistry {
     }
 
     /**
-    * @dev should be impossible to renounce ownership;
+     * @dev should be impossible to renounce ownership;
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/TokenRegistry.sol:414]: format-only change

File: packages/contracts-bridge/contracts/test/TestBridgeMessage.sol [hunk 9]
```diff
@@ -85,10 +85,7 @@ contract TestBridgeMessage {
         uint256 _amnt,
         bytes32 _detailsHash
     ) external view returns (bytes memory) {
-        return
-            BridgeMessage
-                .formatTransfer(_to, _amnt, _detailsHash)
-                .clone();
+        return BridgeMessage.formatTransfer(_to, _amnt, _detailsHash).clone();
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/test/TestBridgeMessage.sol:88]: format-only change

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 10]
```diff
@@ -225,6 +244,72 @@ contract BridgeRouter is Version0, Router {
...
+     * @notice Debit tokens from the function caller
+     * as part of sending tokens across chains
+     * @dev Locks canonical tokens in escrow in BridgeRouter
+     * OR Burns representation tokens
+     * @param _token The token to pull from the sender
+     * @param _amount The amount to pull from the sender
+     * @return _tokenId the bytes canonical token identifier
+     * @return _detailsHash the hash of the canonical token details (name, symbol, decimal)
+     */
+    function _debitTokens(address _token, uint256 _amount)
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:259]: this is a non-functional refactor, and should be in a separate PR before this one.

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 11]
```diff
@@ -244,16 +329,77 @@ contract BridgeRouter is Version0, Router {
...
+     * @param _recipient The recipient that will receive tokens
+     * @return _token The address of the local token contract
+     */
+    function _creditTokens(
+        uint32 _origin,
+        uint32 _nonce,
+        bytes29 _tokenId,
+        bytes29 _action,
+        address _recipient
+    ) internal returns (address _token) {
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:394]: this is a non-functional refactor, and should be in a separate PR before this one.

Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeRouter.sol:394]: (saying I approve of the changes, but they make the functional changes harder to review)

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 12]
```diff
@@ -225,6 +244,72 @@ contract BridgeRouter is Version0, Router {
...
+    /**
+     * @notice Dispatch a message via Nomad to a destination domain
+     * addressed to the remote BridgeRouter on that chain
+     * @dev Message will trigger `handle` method on the remote BridgeRouter
+     * when it is received on the destination chain
+     * @param _destination The domain of the destination chain
+     * @param _tokenId The canonical token identifier for the transfer message
+     * @param _action The contents of the transfer message
+     */
+    function _sendTransferMessage(
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:298]: non-functional change should be in a separate PR

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 13]
```diff
@@ -244,16 +329,77 @@ contract BridgeRouter is Version0, Router {
...
+        // send tokens
+        address _token = _creditTokens(
+            _origin,
+            _nonce,
+            _tokenId,
+            _action,
+            _recipient
+        );
+        // call Connext reconcile callback
+        connext.reconcile(
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:368]: this call will potentially revert. This can cause user funds to become permanently unretrievable

**Comment by anna-carroll** [comment]
> To make review easier, please remove format-only changes to `BridgeToken`, `TokenRegistry`, and `BridgeMessage`. What caused these? These changes aren't made with a local prettier run

They were caused by running prettier on my machine

File: packages/contracts-bridge/contracts/BridgeRouter.sol (hunk 10 above)
Comment (anna-carroll) [reply packages/contracts-bridge/contracts/BridgeRouter.sol]: https://github.com/nomad-xyz/monorepo/pull/385

File: packages/contracts-bridge/contracts/BridgeRouter.sol (hunk 11 above)
Comment (anna-carroll) [reply packages/contracts-bridge/contracts/BridgeRouter.sol]: https://github.com/nomad-xyz/monorepo/pull/385

File: packages/contracts-bridge/contracts/BridgeRouter.sol (hunk 12 above)
Comment (anna-carroll) [reply packages/contracts-bridge/contracts/BridgeRouter.sol]: https://github.com/nomad-xyz/monorepo/pull/385

File: packages/contracts-bridge/contracts/BridgeMessage.sol (hunk 3 above)
Comment (anna-carroll) [reply packages/contracts-bridge/contracts/BridgeMessage.sol:24]: can you explain further what you mean?

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 14]
```diff
@@ -155,6 +170,41 @@ contract BridgeRouter is Version0, Router {
...
+     * @param _externalId A hash of transfer metadata for Connext flow
+     */
+    function xsend(
+        address _token,
+        uint256 _amount,
+        uint32 _destination,
+        bytes32 _externalId
+    ) external {
+        // validate inputs
+        require(_externalId != bytes32(0), "!id");
```
Comment (odyslam) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:187]: An idea could be to move away from input validation, which increases the net gas costs for everyone. I would expect such checks to be conducted on the front end, as smart contracts should not be used directly.

File: packages/contracts-bridge/CHANGELOG.md [hunk 15]
```diff
@@ -2,6 +2,7 @@
 
 ### Unreleased
 - remove ability to handle fast liquidity messages
+- refactor token transfer path with readable helper functions
```
Comment (anna-carroll) [inline packages/contracts-bridge/CHANGELOG.md:5]: todo: update changelog

File: packages/contracts-bridge/contracts/BridgeMessage.sol [hunk 16]
```diff
@@ -326,6 +364,50 @@ library BridgeMessage {
...
+    {
+        return _transferAction.indexAddress(13);
+    }
+
+    /**
+     * @notice Retrieves the hook contract from a TransferWithHook
+     * @param _transferAction The message
+     * @return The hook contract address as an address
+     */
+    function hook(bytes29 _transferAction)
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeMessage.sol:386]: is this ever needed?

File: packages/contracts-bridge/contracts/BridgeMessage.sol [hunk 17]
```diff
@@ -326,6 +364,50 @@ library BridgeMessage {
...
+        pure
+        typeAssert(_transferAction, Types.TransferToHook)
+        returns (bytes29)
+    {
+        // anything past the end is the extradata
+        return
+            _transferAction.slice(
+                TRANSFER_LEN,
+                _transferAction.length - TRANSFER_LEN,
+                uint40(Types.ExtraData)
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeMessage.sol:407]: why have a type for this?

File: packages/contracts-bridge/contracts/BridgeMessage.sol [hunk 18]
```diff
@@ -59,7 +61,7 @@ library BridgeMessage {
      * @return TRUE if action is valid
      */
     function isValidAction(bytes29 _action) internal pure returns (bool) {
-        return isTransfer(_action);
+        return isTransfer(_action) || isTransferToHook(_action);
     }
 
     /**
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeMessage.sol:67]: TODO: update isValidMessageLength - expects exactly 
```
return _len == TOKEN_ID_LEN + TRANSFER_LEN;
```

File: packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+interface IBridgeHook {
+    /**
+     * @notice Handles an incoming bridge transfer with some tokens and extra
+     * data. Takes any necessary actions for the hook's purposes.
+     *
+     * This function is intended to allow arbitrary post-bridge actions with
+     * tokens, at a user's discretion. E.g. recollateralize a CDP, exchange for
+     * other tokens, emit an event, etc.
+     *
+     * This hook is called AFTER tokens have been transferred to the hook
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol:13]: todo: add WARNING

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 20]
```diff
@@ -136,49 +139,102 @@ contract BridgeRouter is Version0, Router {
...
+    /**
+     * @notice Send tokens to a hook on the remote chain
+     * @param _token The token address
+     * @param _amount The token amount
+     * @param _destination The destination domain
+     * @param _hook The hook contract on the remote chain
+     * @param _extraData Extra data that will be passed to the hook for
+     *        execution
+     */
+    function xsend(
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:203]: typechain will be ugly if we have 2 functions with the same name. we should probably only have the `bytes32` version

File: packages/contracts-bridge/contracts/BridgeRouter.sol (hunk 13 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeRouter.sol]: update: we have checked this with connext. we're going to make this call ExcessivelySafe

File: packages/contracts-bridge/contracts/BridgeMessage.sol (hunk 16 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeMessage.sol:386]: not really. deleted

File: packages/contracts-bridge/contracts/BridgeMessage.sol (hunk 17 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeMessage.sol:407]: cleanliness, basically.

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 21]
```diff
@@ -40,7 +41,7 @@ contract BridgeRouter is Version0, Router {
     // ============ Upgrade Gap ============
 
     // gap for upgrade safety
-    uint256[49] private __GAP;
+    uint256[48] private __GAP;
```
[the remark this answers is not in the document]
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeRouter.sol:44]: 👍

File: packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol (hunk 19 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol:13]: add warning to what?

File: packages/contracts-bridge/contracts/BridgeMessage.sol [hunk 22]
```diff
@@ -326,6 +363,35 @@ library BridgeMessage {
...
+        return _transferAction.indexAddress(13);
+    }
+
+    function extraData(bytes29 _transferAction)
+        internal
+        pure
+        typeAssert(_transferAction, Types.TransferToHook)
+        returns (bytes29)
+    {
+        // anything past the end is the extradata
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeMessage.sol:386]: thought: what if the action is equal to `MIN_TRANSFER_LEN`? does `slice` behave as expected if the number of bytes to slice is 0

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 23]
```diff
@@ -155,6 +158,39 @@ contract BridgeRouter is Version0, Router {
...
+     *        execution
+     */
+    function xsend(
+        address _token,
+        uint256 _amount,
+        uint32 _destination,
+        bytes32 _hook,
+        bytes calldata _extraData
+    ) external {
+        // debit tokens from the Connext
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:177]: todo: remove "connext" from comment

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 24]
```diff
@@ -155,6 +158,39 @@ contract BridgeRouter is Version0, Router {
...
+    /**
+     * @notice Send tokens to a hook on the remote chain
+     * @param _token The token address
+     * @param _amount The token amount
+     * @param _destination The destination domain
+     * @param _hook The hook contract on the remote chain
+     * @param _extraData Extra data that will be passed to the hook for
+     *        execution
+     */
+    function xsend(
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:170]: we should probably rename this function? `sendToHook` ?

Comment (anna-carroll) [reply packages/contracts-bridge/contracts/BridgeRouter.sol:170]: another naming note: not clearest naming to me that `hook` connotes a contract which implements the logic for a hook

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 25]
```diff
@@ -155,6 +158,39 @@ contract BridgeRouter is Version0, Router {
...
+        address _token,
+        uint256 _amount,
+        uint32 _destination,
+        bytes32 _hook,
+        bytes calldata _extraData
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:175]: if `_extraData` is empty, does everything still function as normal?

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 26]
```diff
@@ -287,6 +323,45 @@ contract BridgeRouter is Version0, Router {
         _dust(_recipient);
     }
 
+    /**
+     * @notice Handles an incoming TransferToHook message.
+     *
+     * Tokens are sent to the Hook contract
+     * then IBridgeHook.onReceive callback is called
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:330]: add
@dev for behavior if `IBridgeHook.onReceive` reverts

File: packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol [hunk 27]
```diff
@@ -0,0 +1,35 @@
...
+     * @param _localToken The local address of the token (its canonical
+     *                    address if it is local to this domain, otherwise its
+     *                    the address of its local representation).
+     * @param _amount The amount of token received over the bridge
+     * @param _extraData Extra user-specified data passed in to the origin chain
+     */
+    function onReceive(
+        uint32 _origin,
+        uint32 _tokenDomain,
+        bytes32 _tokenAddress,
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol:30]: should we rename to `_tokenIdentifier`?

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 28]
```diff
@@ -287,6 +323,45 @@ contract BridgeRouter is Version0, Router {
...
+            _hook
+        );
+        bytes memory _call = abi.encodeWithSelector(
+            IBridgeHook.onReceive.selector,
+            _origin,
+            _tokenId.domain(),
+            _tokenId.id(),
+            _token,
+            _action.amnt(),
+            _action.extraData().clone()
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:360]: does this work if `extraData` is empty?

**Review by anna-carroll — COMMENTED** [review]
🌊

File: packages/contracts-bridge/contracts/BridgeMessage.sol (hunk 17 above)
Comment (anna-carroll) [reply packages/contracts-bridge/contracts/BridgeMessage.sol:391]: i don't think i understand what's more clean about this

File: packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol (hunk 19 above)
Comment (anna-carroll) [reply packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol:13]: sorry this was my brain speaking to itself. 

was thinking it would be prudent to prefix this comment with `WARNING:` or `NOTICE:` or something to draw attention to this for developers

**Comment by anna-carroll** [comment]
We need to update documentation to reflect the inclusion of this feature

File: packages/contracts-bridge/contracts/BridgeMessage.sol (hunk 22 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeMessage.sol:386]:
```
function slice(bytes29 memView, uint256 _index, uint256 _len, uint40 newType) internal pure returns (bytes29) {
        uint256 _loc = loc(memView);

        // Ensure it doesn't overrun the view
        if (_loc + _index + _len > end(memView)) {
            return NULL;
        }

        _loc = _loc + _index;
        return build(newType, _loc, _len);
    }
```

it'll return a slice of 0 length

File: packages/contracts-bridge/contracts/BridgeRouter.sol (hunk 28 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeRouter.sol:360]: yes, clone will produce a valid, empty, bytes mem

**Comment by prestwich** [comment]
> We need to update documentation to reflect the inclusion of this feature

issue on docs repo?

File: packages/contracts-bridge/contracts/BridgeRouter.sol (hunk 25 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeRouter.sol:175]: cc @odyslam

File: packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol (hunk 19 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/interfaces/IBridgeHook.sol]: think the caps do that pretty well already

File: packages/contracts-bridge/contracts/BridgeRouter.sol (hunk 13 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeRouter.sol]: using `0` for the type everywhere is basically unintended behavior for the lib. it's cleaner to type your typed mem views

File: packages/contracts-bridge/contracts/BridgeMessage.sol (hunk 17 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/BridgeMessage.sol:391]: using 0 for the type everywhere is basically unintended behavior for the lib. it's cleaner to type your typed mem views

**Comment by anna-carroll** [comment]
> > We need to update documentation to reflect the inclusion of this feature
> 
> issue on docs repo?

i think the docs repo is no longer the source of changes for docs - docs now live in gitbook. will escalate convo to slack

File: packages/contracts-bridge/contracts/BridgeMessage.sol [hunk 29]
```diff
@@ -132,12 +143,37 @@ library BridgeMessage {
...
         bytes32 _detailsHash
+    ) internal pure returns (bytes29) {
+        return
+            abi.encodePacked(Types.Transfer, _to, _amnt, _detailsHash).ref(
+                uint40(Types.Transfer)
+            );
+    }
+
+    /**
+     * @notice Formats Connext Transfer
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeMessage.sol:154]: remove Connext from comment

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 30]
```diff
@@ -155,6 +158,39 @@ contract BridgeRouter is Version0, Router {
...
+     *        execution
+     */
+    function sendToHook(
+        address _token,
+        uint256 _amount,
+        uint32 _destination,
+        bytes32 _remoteHook,
+        bytes calldata _extraData
+    ) external {
+        // debit tokens from the Hook
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:177]: update: `debit tokens from msg.sender`

File: packages/contracts-bridge/contracts/BridgeRouter.sol [hunk 31]
```diff
@@ -282,23 +326,65 @@ contract BridgeRouter is Version0, Router {
...
+        address _hook = _action.evmHook();
+        // send tokens
+        address _token = _giveTokens(
+            _origin,
+            _nonce,
+            _tokenId,
+            _action,
+            _hook
+        );
+        bytes memory _call = abi.encodeWithSelector(
```
Comment (anna-carroll) [inline packages/contracts-bridge/contracts/BridgeRouter.sol:363]: add comment above this line

**Review by anna-carroll — COMMENTED** [review]
amazing work!! LGTM! ✅

**Review by prestwich — APPROVED** [review]
stop. Collaborate. And Listen.
